File: day02/inventory.py
#!/usr/bin/env python3
from sys import argv
from collections import defaultdict

def count_appear(s):
    d = defaultdict(int)
    for c in s:
        d[c] += 1

    has2 = False
    has3 = False
    for _, v in d.items():
        if v == 2:
            has2 = True
        elif v == 3:
            has3 = True

    return has2, has3

def letters_in_common(s, t):
    res = ''
    for i in range(len(s)):
        if s[i] == t[i]:
            res += s[i]
    return res

num2 = 0
num3 = 0
ids = []
filename = argv[1]
with open(filename) as f:
    for line in f:
        id = line.rstrip()
        ids.append(id)
        has2, has3 = count_appear(id)
        if has2:
            num2 += 1
        if has3:
            num3 += 1

print(f'checksum = {num2 * num3}')

target_len = len(ids[0]) - 1
for i in range(len(ids)-1):
    for j in range(i+1, len(ids)):
        common = letters_in_common(ids[i], ids[j])
        if len(common) == target_len:
            print(common)

File: day02/inventory.pl
#!/usr/bin/env perl
use strict;
use warnings;
use feature qw(:5.24);
use experimental qw(signatures);

my ($num2, $num3) = (0, 0);
my @ids;

while (<>) {
    chomp;
    push @ids, $_;
    my ($has2, $has3) = count_appear($_);
    $num2++ if $has2;
    $num3++ if $has3;
}

say "checksum = ", $num2 * $num3;

my $target_len = length($ids[0]) - 1;
for my $i (0..$#ids - 1) {
    for my $j ($i+1..$#ids) {
	my $common = letters_in_common($ids[$i], $ids[$j]);
	say $common if length($common) == $target_len;
    }
}

sub count_appear($s) {
    my %h;
    for my $c (split //, $s) {
	$h{$c}++;
    }

    my ($has2, $has3) = (0, 0);
    for my $cnt (values %h) {
	$has2 = 1 if $cnt == 2;
	$has3 = 1 if $cnt == 3;
    }

    return ($has2, $has3);
}

sub letters_in_common($s, $t) {
    my $res = '';
    my @s = split //, $s;
    my @t = split //, $t;
    for my $i (0..$#s) {
	$res .= $s[$i] if $s[$i] eq $t[$i];
    }
    return $res;
}
